ravenml/data/helpers.py

```diff
@@ -0,0 +1,311 @@
+                    in that column header
+        filter_tags (list): a list of str values for the tags that should be
+            used to apply the filter
+    
+    Returns:
+        DataFrame: a subset of the input DataFrame with those images that
+            passed the AND filter remaining
+    """
+    subset = tags_df
+    for tag in filter_tags:
+        subset = subset[subset[tag]]
+    return subset
+
+def or_filter(tags_df, filter_tags):
+    """Filters out a set of images based upon the union of its tag values
+    
+    NOTE: With an OR (union) filter, an image must possess at least one tag in
+    the `filter_tags` list in order to pass the filter.
+
+    Args:
+        tags_df (DataFrame): a pandas DataFrame storing image IDs and
+            associated tags; its structure is:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+        filter_tags (list): a list of str values for the tags that should be
+            used to apply the filter
+    
+    Returns:
+        DataFrame: a subset of the input DataFrame with those images that
+            passed the OR filter remaining
+    """
+    result = pd.DataFrame()
+    for tag in filter_tags:
+        subset = tags_df[tags_df[tag]]
+        result = pd.concat((result, subset), sort=False)
+    result = result.reset_index().drop_duplicates(
+        subset='index', keep='first').set_index('index')
+    return result
+
+def join_sets(sets):
+    """Returns the union of a set of datasets
+    
+    NOTE: this function removes duplicates, so it is possible to end up with a
+    smaller number of items than are given as inputs.
+
+    Args:
+        sets (list): a list of datasets that should be merged into one (the
+            union of all datasets); each element of the list is a DataFrame
+            in the format:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+    
+    Returns:
+        DataFrame: a DataFrame that represents the union of all of the input
+            DataFrames with duplicates removed; it stores image IDs and 
+            associated tags in the format:
+                index (rows) = image ID (str)
+                column headers = the tags themselves
+                columns = True/False values for whether the image has the tag
+                    in that column header
+    """
+    result = pd.DataFrame()
+    for group in sets:
+        result = pd.concat((result, group), sort=False)
+    result = result.reset_index().drop_duplicates(
+        subset='index', keep='first').set_index('index')
+    return result
+
+def copy_data_locally(images: list, destination_dir: Path, associated_files: dict, num_threads=20):
```

Comment:
This name is kind of a misnomer, since it doesn't have anything to do with "local" data anymore (I know this is a holdover from Jigsaw when the temp dir was always in the CWD). Perhaps `copy_associated_files` would be a better name.